packages/light-client-extension-helpers/src/background/json-rpc-provider.ts

```diff
@@ -0,0 +1,205 @@
+     * It is the responsibility of the JSON RPC provider to consume these
+     * messages and propagate them with the onMessage callback.
+     */
+    const messagePubSub = yield* PubSub.unbounded<string>()
+
+    /**
+     * An error pubsub for the daemon to publish errors to.
+     * It is the responsibility of the JSON RPC provider to consume these
+     *  errors and propagate them with the onError callback.
+     */
+    const errorPubSub = yield* PubSub.sliding<void>(1)
+
+    /**
+     * A synchronized reference to the current chain. This reference is updated
+     * every time the daemon restarts.
+     *
+     */
+    const chainRef = yield* $(
+      Effect.tryPromise(() => client.addChain(options)),
+      Effect.andThen(SynchronizedRef.make),
+    )
+
+    /**
+     * This daemon process listens for JSON RPC messages, publishing them to
+     * the message pubsub. It also detects errors and publishes those to the
+     * error pubsub.
+     *
+     * If an error occurs with `chain.nextJsonRpcResponse()`, the daemon
+     * restarts. This indicates that the chain is no longer functional.
+     * In response, the daemon reconnects to smoldot using the same
+     * `AddChainParameters` to obtain a new `Chain` object.
+     */
+    const daemon = yield* Effect.gen(function* () {
+      yield* Effect.addFinalizer(() => Queue.takeAll(daemonReadySignal))
+      yield* Effect.addFinalizer(() => PubSub.publish(errorPubSub, undefined))
+
+      const chain = yield* SynchronizedRef.updateAndGetEffect(
+        chainRef,
+        (oldChain) =>
+          $(
+            Effect.tryPromise(() => client.addChain(options)),
+            Effect.tap(() =>
+              $(
+                Effect.try(() => oldChain.remove()),
+                Effect.catchAll(() => Effect.void),
+              ),
+            ),
+          ),
+      )
+
+      // chain is initialized so we can signal that the daemon is ready
+      yield* Queue.offer(daemonReadySignal, undefined)
+
+      yield* $(
+        Effect.tryPromise(() => chain.nextJsonRpcResponse()),
+        Effect.tap(Effect.logDebug),
+        Effect.tap((message) => Queue.offer(messagePubSub, message)),
+      ).pipe(Effect.forever)
+    }).pipe(
+      // by scoping here, we ensure the finalizers are run before the next retry
+      Effect.scoped,
+      Effect.retry(Schedule.spaced("1 second")),
+      Effect.forkDaemon,
+      Effect.interruptible,
+    )
+
+    const provider: JsonRpcProvider = yield* Effect.sync(() =>
+      getSyncProvider(async () => {
+        await $(
+          Queue.take(daemonReadySignal),
+          Effect.tap(() => Effect.log("JSON RPC provider initialized.")),
+          runPromise,
+        )
+
+        return (onMessage, onError) => {
+          const invokeOnMessageCallback = (message: string) =>
+            $(
+              Effect.try(() => onMessage(message)),
+              Effect.catchAll(() => Effect.void),
+            )
```

Comment:
This is basically just the following, right?

```ts
try {
    onMessage(message)
} catch {
    // Ignore error calling onMessage
}
```

Reply:
I think this works and is more concise

```ts
return (_onMessage, _onError) => {
  const onMessage = (message: string) =>
    Effect.try({
      try: () => _onMessage(message),
      catch: () => {},
    })

  const onError = Effect.try({ try: _onError, catch: () => {} })
```